src/process_impl/mod.rs

```diff
@@ -0,0 +1,203 @@
+        if self.terminated {
+            return Err(Error::Disconnected);
+        }
+
+        let check = process::Command::new("ssh")
+            .arg("-S")
+            .arg(self.ctl_path())
+            .arg("-o")
+            .arg("BatchMode=yes")
+            .arg("-O")
+            .arg("check")
+            .arg(&self.addr)
+            .output()
+            .await
+            .map_err(Error::Ssh)?;
+
+        if let Some(255) = check.status.code() {
+            if let Some(master_error) = self.take_master_error().await {
+                Err(master_error)
+            } else {
+                Err(Error::Disconnected)
+            }
+        } else {
+            Ok(())
+        }
+    }
+
+    pub(crate) fn raw_command<S: AsRef<OsStr>>(&self, program: S) -> Command {
+        // XXX: Should we do a self.check() here first?
+
+        // NOTE: we pass -p 9 nine here (the "discard" port) to ensure that ssh does not
+        // succeed in establishing a _new_ connection if the master connection has failed.
+
+        let mut cmd = process::Command::new("ssh");
+        cmd.arg("-S")
+            .arg(self.ctl_path())
+            .arg("-T")
+            .arg("-o")
+            .arg("BatchMode=yes")
+            .arg("-p")
+            .arg("9")
+            .arg(&self.addr)
+            .arg("--")
+            .arg(program);
+
+        Command::new(cmd)
+    }
+
+    pub(crate) async fn request_port_forward(
+        &self,
+        forward_type: ForwardType,
+        listen_socket: &Socket<'_>,
+        connect_socket: &Socket<'_>,
+    ) -> Result<(), Error> {
+        let flag = match forward_type {
+            ForwardType::Local => "-L",
+            ForwardType::Remote => "-R",
+        };
+
+        process::Command::new("ssh")
+            .arg("-S")
+            .arg(self.ctl_path())
+            .arg("-o")
+            .arg("BatchMode=yes")
+            .arg("-fNT")
+            .arg(flag)
+            .arg(&format!("{}:{}", listen_socket, connect_socket))
+            .arg(&self.addr)
+            .output()
+            .await
+            .map_err(Error::Ssh)?;
+
+        Ok(())
+    }
+
+    pub(crate) async fn close(mut self) -> Result<(), Error> {
+        self.terminate().await
+    }
+
+    async fn take_master_error(&self) -> Option<Error> {
+        let (_stdout, mut stderr) = self.master.lock().unwrap().take()?;
+
+        let mut err = String::new();
+        if let Err(e) = stderr.read_to_string(&mut err).await {
+            return Some(Error::Master(e));
+        }
+        let stderr = err.trim();
+
+        if stderr.is_empty() {
+            return None;
+        }
+
+        let kind = if stderr.contains("Connection to") && stderr.contains("closed by remote host") {
+            io::ErrorKind::ConnectionAborted
+        } else {
+            io::ErrorKind::Other
+        };
+
+        Some(Error::Master(io::Error::new(kind, stderr)))
+    }
+
+    async fn terminate(&mut self) -> Result<(), Error> {
```

Comment:
I _think_ we can just move this into `close` now. I originally kept it separate and gave it only `&mut self` so that it could be called from `drop`, but then changed `drop` to use _synchronous_ `std::process::Command`, so that's no longer helpful.

Reply:
Done.